### DISCUSSION TITLE
Standardizing integration tests for popular dbt packages

### DISCUSSION THREAD

**Discussion by emmyoop** [dbt-labs/dbt-core#10668] [Community]
dbt Labs has begun the process of testing compatibility of popular dbt packages while testing & releasing changes to dbt in dbt Cloud. 

Our goal is to ensure that dbt Cloud customers can confidently receive continuous dbt updates (”Versionless”), while also enabling us to proactively catch any incompatibilities or functional regressions before they are released in final versions of dbt Core OSS. Our focus to start is on packages that are used by a large number of projects, including packages maintained by dbt Labs and external maintainers. To that end, we are asking package maintainers to standardize a few details of their integration testing setup.

## Target packages

We will start by testing the packages in the [Package hub](https://hub.getdbt.com/) with the greatest usage. Currently, this includes:

### *dbt Labs maintained packages:*

- ✅ [dbt_utils](https://hub.getdbt.com/dbt-labs/dbt_utils/latest/)
- ✅ [audit_helper](https://hub.getdbt.com/dbt-labs/audit_helper/latest/)
- ✅ [codegen](https://hub.getdbt.com/dbt-labs/codegen/latest/)
- ✅ [dbt_external_tables](https://hub.getdbt.com/dbt-labs/dbt_external_tables/latest/)
- ✅ [dbt_project_evaluator](https://hub.getdbt.com/dbt-labs/dbt_project_evaluator/latest/)
 
### *Externally maintained packages, in alphabetical order:*

- [automate_dv](https://hub.getdbt.com/Datavault-UK/automate_dv/latest/)
- [dbt_artifacts](https://github.com/godatadriven/dbt-date)
- [dbt_date](https://hub.getdbt.com/calogica/dbt_date/latest/)
- ✅ [dbt_expectations](https://hub.getdbt.com/metaplane/dbt_expectations/latest/)
- [tsql_utils](https://hub.getdbt.com/dbt-msft/tsql_utils/latest/)
- [elementary](https://hub.getdbt.com/elementary-data/elementary/latest/)
- [fivetran_utils](https://hub.getdbt.com/fivetran/fivetran_utils/latest/)
- [dbt_snowflake_query_tags](https://hub.getdbt.com/get-select/dbt_snowflake_query_tags/latest/)
- [dbt_meta_testing](https://hub.getdbt.com/tnightengale/dbt_meta_testing/latest/)

To support this testing, package maintainers — the experts on the package functionality — will be the ones to maintain their suite of tests. (Most popular packages already do; thank you!) To ensure those tests will run during our automated CI/CD process, dbt Labs will expect a few standard elements of the testing setup. Note that the specific tests which run — whether dbt commands, pytest, or something else — are ultimately up to the package maintainer.

## Requirements for package maintainers

1. Tests will live in `/integration_tests`
2. Supported adapters will be defined in `supported_adapters.env` in the project root.  
<details><summary>
    Example
</summary>

The variable will be called `SUPPORTED_ADAPTERS` and the supported adapters will be listed after it, comma separated.  Note that we will only support adapters running in dbt Cloud.

```
SUPPORTED_ADAPTERS=postgres,snowflake,redshift,bigquery
```

</details>

3. [tox](https://tox.wiki/en/latest/config.html) is used for running tests
4. Each adapter test will be defined in a specific way in the `tox.ini` file. 

<details><summary>
    Example
</summary>

The naming pattern is `testenv:dbt_integration_<ADAPTER>`.

```
    [testenv:dbt_integration_postgres]
    changedir = integration_tests
    allowlist_externals = 
        dbt
    skip_install = true
    commands =
        dbt --version
        dbt debug --target postgres
        dbt deps --target postgres
        dbt build --target postgres --full-refresh
```

</details>

5. The `profiles.yml` used for integration tests will use specific naming conventions for the target, as well as environment variables, so that we can fill the values in our testing process.

  <details><summary>
    Example
</summary>

```yaml
    integration_tests:
      target: postgres
      outputs:
        postgres:
          type: postgres
          host: "{{ env_var('POSTGRES_HOST') }}"
          user: "{{ env_var('POSTGRES_USER') }}"
          pass: "{{ env_var('DBT_ENV_SECRET_POSTGRES_PASS') }}"
          port: "{{ env_var('POSTGRES_PORT') | as_number }}"
          dbname: "{{ env_var('POSTGRES_DATABASE') }}"
          schema: "{{ env_var('POSTGRES_SCHEMA') }}"
          threads: 5
```

</details>  

##  Examples

We've already implemented this setup for `dbt-utils` and are running these tests regularly. To see the setup in action, check out the merged PR [#919](https://github.com/dbt-labs/dbt-utils/pull/919).

For specific requirements, explore our sample package set up for all supported adapters: [dbt-package-testing](https://github.com/dbt-labs/dbt-package-testing).

## Good to Know

We recommend running your own CI using these same tox commands. This way, you can monitor that changes will pass reliably in your CI to be more confident they will also pass in our CI.

**Comment by emmyoop** [comment]
dbt Labs has opened issues in all relevant package repositories.

| package | issue | pr |
|--------|--------|--------|
| [dbt_utils](https://hub.getdbt.com/dbt-labs/dbt_utils/latest/) | https://github.com/dbt-labs/dbt-utils/issues/948 | ✔️ https://github.com/dbt-labs/dbt-utils/pull/919 |
| [audit_helper](https://hub.getdbt.com/dbt-labs/audit_helper/latest/) | https://github.com/dbt-labs/dbt-audit-helper/issues/109 | ✔️https://github.com/dbt-labs/dbt-audit-helper/pull/111 |
| [codegen](https://hub.getdbt.com/dbt-labs/codegen/latest/) | https://github.com/dbt-labs/dbt-codegen/issues/183 | ✔️ https://github.com/dbt-labs/dbt-codegen/pull/181 |
| [dbt_external_tables](https://hub.getdbt.com/dbt-labs/dbt_external_tables/latest/) | https://github.com/dbt-labs/dbt-external-tables/issues/316 | ✔️https://github.com/dbt-labs/dbt-external-tables/pull/320 |
| [dbt_project_evaluator](https://hub.getdbt.com/dbt-labs/dbt_project_evaluator/latest/) | https://github.com/dbt-labs/dbt-project-evaluator/issues/491 | ✔️ https://github.com/dbt-labs/dbt-project-evaluator/pull/512 |
| [automate_dv](https://hub.getdbt.com/Datavault-UK/automate_dv/latest/) | https://github.com/Datavault-UK/automate-dv/issues/248 | |
| [dbt_artifacts](https://hub.getdbt.com/brooklyn-data/dbt_artifacts/latest/) | https://github.com/brooklyn-data/dbt_artifacts/issues/443 | https://github.com/brooklyn-data/dbt_artifacts/pull/451 |
| [~~dbt_date~~](https://hub.getdbt.com/calogica/dbt_date/latest/) | ~~https://github.com/calogica/dbt-date/issues/131~~ | ~~https://github.com/calogica/dbt-date/pull/132~~ |
| [godatadriven/dbt-date](https://github.com/godatadriven/dbt-date) |  |  |
| [~~dbt_expectations~~](https://hub.getdbt.com/calogica/dbt_expectations/latest/) | ~~https://github.com/calogica/dbt-expectations/issues/318~~ | ~~https://github.com/calogica/dbt-expectations/pull/319~~ |
| [dbt_expectations](https://hub.getdbt.com/metaplane/dbt_expectations/latest/) |  https://github.com/metaplane/dbt-expectations/issues/9 | ✔️ https://github.com/metaplane/dbt-expectations/pull/8 |
| [tsql_utils](https://hub.getdbt.com/dbt-msft/tsql_utils/latest/) | https://github.com/dbt-msft/tsql-utils/issues/105 | |
| [elementary](https://hub.getdbt.com/elementary-data/elementary/latest/) | https://github.com/elementary-data/elementary/issues/1700 | https://github.com/elementary-data/elementary/pull/1721 |
| [fivetran_utils](https://hub.getdbt.com/fivetran/fivetran_utils/latest/) | https://github.com/fivetran/dbt_fivetran_utils/issues/149 | |
| [dbt_snowflake_query_tags](https://hub.getdbt.com/get-select/dbt_snowflake_query_tags/latest/) | https://github.com/get-select/dbt-snowflake-query-tags/issues/32 | https://github.com/get-select/dbt-snowflake-query-tags/pull/33 |
| [dbt_meta_testing](https://hub.getdbt.com/tnightengale/dbt_meta_testing/latest/) | https://github.com/tnightengale/dbt-meta-testing/issues/31 | |

**Comment by algarbar** [comment]
@emmyoop -  This is awesome! Please consider adding [dbt_assertions](https://hub.getdbt.com/AxelThevenot/dbt_assertions/latest/).

(@AxelThevenot for visibility)

**Comment by ammarsaf** [comment]
Is there any of this package that has something that can test the data type?

**Comment by pgoslatara** [comment]
@emmyoop Calogica's [dbt-date](https://github.com/calogica/dbt-date) package is no longer supported (similar to their [dbt-expectations](https://github.com/calogica/dbt-expectations) package). At Xebia Data we are partners with dbt Labs and are willing to maintain this package via [godatadriven/dbt-date](https://github.com/godatadriven/dbt-date) and have opened a [PR](https://github.com/dbt-labs/hubcap/pull/355) to add this to the dbt Hub. 

Can you update your list to reflect this? I'm happy to be the point of contact if you have any questions on this.

**Reply by emmyoop** [reply]
I've updated our testing infra to include this ownership change.  In order to be included in the testing you'll need to implement some changes similar to what we had proposed for the `calogica` version: https://github.com/calogica/dbt-expectations/pull/319

Once you've added this testing pattern and released it, it will get pulled in to our testing processes automatically.

**Reply by pgoslatara** [reply]
@emmyoop Done in [0.12.0](https://github.com/godatadriven/dbt-date/releases/tag/0.12.0). Let me know if there's anything else I should do.

Do I have any visibility on what testing processes will be run?

**Reply by emmyoop** [reply]
@pgoslatara our testing resides in an internal repository so you cannot see it.  However, I have modeled the recommend CI you use, [run_tox.yml](https://github.com/dbt-labs/dbt-package-testing/blob/main/.github/workflows/run_tox.yml) to mirror what we're doing for a single package.

Here's the basic internal flow we're using, if you're curious:
1. Build an updated package of core + adapters
2. Download each pinned package listed above from the hub.  The version pins are moved once per month.  I'm moving them today in fact!
3. Check if the package is set up as expected with tox, `supported_adapters.env` etc, as documented above
    a. If it's not set up, I do not error, I simply move on to the next package
    b. If it is set up correctly, I will them use the tox commands you have defined and run the tests against all adapters you define as supported against our new core + adapters package

Upon **success** of all packages (along with many other tests) we will release the new image for dbt Cloud consumption

If there is a **failure** in any of the package tests it will block the release of the updated image and we will look into what the issue is.  If we determine it's the result of a change on our end, we will resolve it.  If it is the result of a change on your end we will reach out.  

Hope that helps!

**Reply by pgoslatara** [reply]
That's really helpful, thanks @emmyoop! I have been looking at the `dbt-package-testing` repo as I built more CI into `dbt-date`, there are a few rough edges that I will try to open PRs against so that others can benefit.
